### PR TITLE
Create ionsmasher.py

### PR DESCRIPTION
**Pull request by meramsey**
Add new variant with:
Config profile or environmental variables support so passwords don't need to be explicitly put on command line each time and in history.

Configurable logging so that you don't lose what happened if your terminal doesn't have infinite scrollback to find and review stuff.

Also is more specific in searching only for files that have `if(!extension_loaded('ionCube Loader'))` vs just `ionCube Loader` which means less false positives on failed decryption which can happen from files with that string but aren't encrypted.
https://stackoverflow.com/questions/2732094/find-ioncube-php-files

Real example:
```
if(!extension_loaded('ionCube Loader')){$__oc=strtolower(substr(php_uname(),0,3));$__ln='ioncube_loader_'.$__oc.'_'.substr(phpversion(),0,3).(($__oc=='win')?'.dll':'.so');if(function_exists('dl')){@dl($__ln);}if(function_exists('_il_exec')){return _il_exec();}$__ln='/ioncube/'.$__ln;$__oid=$__id=realpath(ini_get('extension_dir'));$__here=dirname(__FILE__);if(strlen($__id)>1&&$__id[1]==':'){$__id=str_replace('\\','/',substr($__id,2));$__here=str_replace('\\','/',substr($__here,2));}$__rd=str_repeat('/..',substr_count($__id,'/')).$__here.'/';$__i=strlen($__rd);while($__i--){if($__rd[$__i]=='/'){$__lp=substr($__rd,0,$__i).$__ln;if(file_exists($__oid.$__lp)){$__ln=$__lp;break;}}}if(function_exists('dl')){@dl($__ln);}}else{die('The file '.__FILE__." is corrupted.\n");}if(function_exists('_il_exec')){return _il_exec();}echo("Site error: the ".(php_sapi_name()=='cli'?'ionCube':'<a href="http://www.ioncube.com">ionCube</a>')." PHP Loader needs to be installed. This is a widely used PHP extension for running ionCube protected PHP code, website security and malware blocking.\n\nPlease visit ".(php_sapi_name()=='cli'?'get-loader.ioncube.com':'<a href="http://get-loader.ioncube.com">get-loader.ioncube.com</a>')." for install assistance.\n\n");exit(199);
?>
```

False positive that would try to decrypt and fail
```
if (stripos($preamble, "extension_loaded('ionCube Loader')") === false)
```
Also modded it so that paths are expanded from ~ so can reference stuff with tildes on OS that support it.

Added spacing in the output and colors for success and failure messages.

Can test it out like this once your config is setup.
```bash
python ionsmasher.py -s ~/Downloads/some/path/to/ioncube_encoded/files -w
```

If this is like feel free to incorporate or replace the main.py contents with this script if desired but wanted it to be easy for you to test locally to see if i missed anything.

Has there been any consideration of adding retry attempts for failed ones? I noticed sometimes on some files first time fails but on subsequent ones that they can succeed. Might be worth adding a retry attempts default like 1-3 to retry before reporting failed at the end.

Also if they fail to decrypt do the files not get copied over? 

If not food for thought:
I think they should be copied but `_unable_to_decode` or similar added maybe vs having to wonder why a file is missing when doing post checks.

### CODE REVIEW

**Comment by meramsey** [comment]
@ip-rw any chance you could review this?

File: ionsmasher.py [hunk 1]
```diff
@@ -0,0 +1,386 @@
...
+
+
+def gray(text):
+    print('\033[90m', text, '\033[0m', sep='')
+
+
+#################################
+parser = argparse.ArgumentParser(usage="easy4us", description="decode directories with easytoyou.eu")
+parser.add_argument("-u", "--username", required=False, help="easytoyou.eu username", default=username)
+parser.add_argument("-p", "--password", required=False, help="easytoyou.eu password", default=password)
```
[the remark this answers is not in the document]
Comment (ip-rw) [reply ionsmasher.py:179]: usernames and passwords are required, i'm not fussed about just returning the standard argparse usage. 

thank you for the PR! honestly it's made my day :) but i'm happy leaving my version as minimal as possible - unpolished as it is

File: ionsmasher.py [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+    if args.username is None:
+        message = 'Please set username in profile config or pass it as an argument via --username'
+        red(message)
+        logging.info(message)
+        sys.exit()
+    if args.password is None:
+        message = 'Please set password in profile config or pass it as an argument via --password'
+        red(message)
+        logging.info(message)
+        sys.exit()
```
[the remark this answers is not in the document]
Comment (ip-rw) [reply ionsmasher.py:361]: same as prev. happy with argparse output - saves loc.

File: ionsmasher.py [hunk 3]
```diff
@@ -0,0 +1,386 @@
...
+            rel = os.path.relpath(dir, args.source)
+            dest = os.path.join(args.destination, rel).strip(".")
+            if not os.path.exists(dest):
+                os.makedirs(dest)
+                # print("created %s" % dest)
+            phpfiles = []
+            other = []
+            for f in filenames:
+                csrc = os.path.join(dir, f)
+                if f.endswith(".php") and b"if(!extension_loaded('ionCube Loader'))" in open(csrc, "rb").read():
```
[the remark this answers is not in the document]
Comment (ip-rw) [reply ionsmasher.py:380]: yep - handy keep that one.

File: ionsmasher.py [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+                    process_files(session, dir, dest, f)
+        green('finished.')
+        if not_decoded:
+            message = "ioncube files that failed to decode:"
+            red(message)
+            logging.info(message)
+            for f in not_decoded:
+                red(f)
+                log_entry = f'failed to decode: {f}'
+                logging.info(log_entry)
```
[the remark this answers is not in the document]
Comment (ip-rw) [reply ionsmasher.py:413]: again its really a matter of personal preference but i like parsable, clean output. horses for courses but again i think better to keep it simple, so others can go their own way with it.

**Review by ip-rw — CHANGES_REQUESTED** [review]
thank you for the PR! honestly it's made my day :) but i'm happy leaving my version as minimal as possible - unpolished as it is

the way you've submitted this means it's not going to merge in any meaningful way - however I think there's some good stuff in here and i really appreciate you making the effort to contribute back. I've noticed at leas one thing i've overlooked and needs a command line arg (copy originals or not, this addresses the retrying thing). 

i'll pick the good bits out by hand if needs be but a mergeable PR would be even better )

File: ionsmasher.py [hunk 5]
```diff
@@ -0,0 +1,413 @@
...
+
+if __name__ == '__main__':
+    # Let's expand paths from ~
+    args.source = os.path.expanduser(args.source)
+    args.destination = os.path.expanduser(args.destination)
+    
+    if args.destination == "":
+        args.destination = os.path.basename(args.source) + "_decoded"
+        # Let's create the decoded folder in same directory as source but with _decoded
+        args.destination = os.path.join(os.path.abspath(os.path.dirname(args.source)), args.destination)
```
[the remark this answers is not in the document]
Comment (ip-rw) [reply ionsmasher.py:344]: you're absolutely right in saying if left blank it should be the cwd. again pls submit or i will change (a default="." will do it)

**Comment by ip-rw** [comment]
> Has there been any consideration of adding retry attempts for failed ones? I noticed sometimes on some files first time fails but on subsequent ones that they can succeed. Might be worth adding a retry attempts default like 1-3 to retry before reporting failed at the end.

i'm going to add a "copy original on fail" switch - it will default to off, so retries should be a matter of just rerunning the script - it won't overwrite existing files in the destination (unless you tell it to)

sometimes there are just files they cant decrypt - then you can copy the originals and ofc the thing still runs.

File: ionsmasher.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ip-rw) [reply ionsmasher.py:344]: >             # Let's replace EasyToYou.eu comment from file content

myself i would leave this - it's a good service, it's beyond the scope of the tool.

**Pull request closed without merging**